expConfig.md
C0:  90 speaker 23400utt[:260:290:320], few_noise, log_mag_norm[-3,6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX
C1:  400 speaker 108000utt[:270:300:320], few_noise, log_mag_norm[-3,6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX
C2:  400 speaker 108000utt[:270:300:320], few_noise, log_mag_norm[-0.3,6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX
C3:  1991 speaker 796400utt[:400:460:520], large_noise, log_mag_norm[-0.3,6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX
C4:  1991 speaker 796400utt[:400:460:520], large_noise, log_mag_norm[-0.3,6], psm, DATASET_SIZES = [600000, 18000, 100000], SNR_MIX
C5:  1991 speaker 796400utt[:400:460:520], large_noise, log_mag_norm[-0.3,6], psm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX
C6:  1991 speaker 796400utt[:400:460:520], large_noise, log_mag_norm[-0.3,6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX, loss1_Wrong,
C7:  1991 speaker 796400utt[:400:460:520], large_noise, mag_norm[0,1e6], irm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX, loss2,[512fft,256hop]，网络输入使用幅度谱-网络输出为幅度谱的mask
C8:  1991 speaker 796400utt[:400:460:520], large_noise, mag_norm[0,1e6], psm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX, loss2,[512fft,256hop]，网络输入使用幅度谱-网络输出为幅度谱的mask
C9:  1991 speaker 796400utt[:400:460:520], large_noise, mag_norm[0,1e6], psm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX, loss1,[1024fft,160hop],网络输入使用幅度谱-网络输出为幅度谱的mask,PESQImp_0.8
C9-2:   loss2
C10: 1991 speaker 796400utt[:400:460:520], large_noise, log_mag_norm[-0.3,6]&mag_norm[0,1e6], psm, DATASET_SIZES = [600000, 18000, 100000], 0-1_MIX, loss2,[1024fft,160hop],网络输入使用对数谱-输出为幅度谱的mask-loss为MSE(估计的对数谱,对数谱)
C11: 网络输入使用对数谱-网络输出为对数谱的mask; log(mag+50/200/1000), 训练时MASK_ON_MAG_EVEN_LOGMAG=None; 解码时MASK_ON_MAG_EVEN_LOGMAG=True,PESQImp_1.0
c11主要有两点，一个是将取对数时的偏移增大，减弱对数谱对噪音的增强；一个是将对数谱的mask直接加到幅度谱上进行解码。


loss1 = loss.reduce_sum_frame_batchsize_MSE
loss2 = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE
loss3 = 联合幅度谱和对数谱





max min avg delta^2 delta
??? 5.9912505 -0.26359642 3.2887695 0.6685478 0.81764776
??? 6.289474 -0.29723358 3.2353096 0.7063788 0.84046346
??? 5.922567 0.34533253 3.191401 0.46772853 0.6839068
??? 6.1767936 -0.066084966 3.049846 0.7746223 0.8801263
??? 6.1386247 -0.17414095 3.0165172 0.5378319 0.73337024

FLAGS.py
from losses import loss

class C3_NNET:
  MASK_TYPE = "IRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 257
  OUTPUT_SIZE = 257
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = True
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C3_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  #RAW_DATA = '/all_data/aishell2_speaker_list' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  #NOISE_DIR = '/all_data/many_noise' # for docker
  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 512
  OVERLAP = NFFT - 256
  FS = 16000
  FEATURE_TYPE = 'LOG_MAG'  # MAG or LOG_MAG
  MASK_ON_MAG_EVEN_LOGMAG = None
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0


  AUDIO_VOLUME_AMP=True


  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/all_data/feature_tfrecords' # for docker
  TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm'

class C6_NNET:
  MASK_TYPE = "IRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 257
  OUTPUT_SIZE = 257
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = True
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C6_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  # RAW_DATA = '/aishell_90_speaker' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 512
  OVERLAP = NFFT - 256
  FS = 16000
  FEATURE_TYPE = 'LOG_MAG'  # MAG or LOG_MAG
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0


  AUDIO_VOLUME_AMP=True


  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/feature_tfrecords_utt03s_irm' # for docker
  TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm'
  if MIX_METHOD =='SNR':
    TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm_SNR_MIX'

class C7_NNET:
  MASK_TYPE = "IRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 257
  OUTPUT_SIZE = 257
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = True
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C7_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  # RAW_DATA = '/aishell_90_speaker' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 512
  OVERLAP = NFFT - 256
  FS = 16000
  FEATURE_TYPE = 'MAG'  # MAG or LOG_MAG
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0


  AUDIO_VOLUME_AMP=True


  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/feature_tfrecords_utt03s_irm' # for docker
  if FEATURE_TYPE == "LOG_MAG":
    TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm'
    if MIX_METHOD =='SNR':
      TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm_SNR_MIX'
  elif FEATURE_TYPE == "MAG":
    TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_mag_utt03s_irm'
    if MIX_METHOD =='SNR':
      TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_mag_utt03s_irm_SNR_MIX'

class C8_NNET:
  MASK_TYPE = "PSIRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 257
  OUTPUT_SIZE = 257
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet_C8'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = False
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C8_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  # RAW_DATA = '/aishell_90_speaker' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 512
  OVERLAP = NFFT - 256
  FS = 16000
  FEATURE_TYPE = 'MAG'  # MAG or LOG_MAG
  MASK_ON_MAG_EVEN_LOGMAG = None # It worked only when FEATURE_TYPE is 'MAG'.
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0

  AUDIO_VOLUME_AMP=False

  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/feature_tfrecords_utt03s_irm' # for docker
  if FEATURE_TYPE == "LOG_MAG":
    TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm'
    if MIX_METHOD =='SNR':
      TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm_SNR_MIX'
  elif FEATURE_TYPE == "MAG":
    TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_mag_utt03s_irm'
    if MIX_METHOD =='SNR':
      TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_mag_utt03s_irm_SNR_MIX'

class C9_NNET:
  MASK_TYPE = "PSIRM"  # or 'PSIRM'
  # LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE # "MSE" "MSE_LOW_FS_IMPROVE"
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 513
  OUTPUT_SIZE = 513
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet_C9-2'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 4
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = False
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C9_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  # RAW_DATA = '/aishell_90_speaker' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 1024
  OVERLAP = 1024-160
  FS = 16000
  FEATURE_TYPE = 'MAG'  # MAG or LOG_MAG
  MASK_ON_MAG_EVEN_LOGMAG = None # It worked only when FEATURE_TYPE is 'MAG'.
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0

  AUDIO_VOLUME_AMP=False

  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/feature_tfrecords_utt03s_irm' # for docker
  TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_mag_utt03s_irm_1024fft'

class C10_NNET:
  MASK_TYPE = "PSIRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 513
  OUTPUT_SIZE = 513
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet'
  '''
  decode:
    decode by the flod '_decode_index'. one set per (.list) file.
  '''
  decode = 1  # 0:train; 1:decode_for_show; 2:decode_test_set_calculate_SDR_Improvement

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = False
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C10_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  # RAW_DATA = '/aishell_90_speaker' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = False
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 1024
  OVERLAP = NFFT - 160
  FS = 16000
  FEATURE_TYPE = 'LOG_MAG'  # MAG or LOG_MAG
  MASK_ON_MAG_EVEN_LOGMAG = True
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 0.5
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0

  AUDIO_VOLUME_AMP=False

  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/feature_tfrecords_utt03s_irm' # for docker
  TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_logmag_utt03s_irm_1024fft'

class C11_NNET:
  MASK_TYPE = "PSIRM"  # or 'PSIRM'
  LOSS_FUNC = loss.reduce_sum_frame_batchsize_MSE_LOW_FS_IMPROVE # "MSE" "MSE_LOW_FS_IMPROVE"
  MODEL_TYPE = 'BLSTM'  # 'BLSTM' or 'BGRU'
  INPUT_SIZE = 257
  OUTPUT_SIZE = 257
  LSTM_num_proj = 128
  RNN_SIZE = 512
  LSTM_ACTIVATION = 'tanh'
  KEEP_PROB = 0.8
  RNN_LAYER = 2
  CLIP_NORM = 5.0
  SAVE_DIR = 'exp/rnn_irm'
  CHECK_POINT = 'nnet_C11_bias50'

  batch_size = 128
  learning_rate = 0.001
  start_halving_impr = 0.0003
  resume_training = 'false'
  start_epoch = 0
  min_epochs = 10  # Min number of epochs to run trainer without halving.
  max_epochs = 50  # Max number of epochs to run trainer totally.
  halving_factor = 0.7  # Factor for halving.
  # Halving when ralative loss is lower than start_halving_impr.
  start_halving_impr = 0.003
  # Stop when relative loss is lower than end_halving_impr.
  end_halving_impr = 0.0005
  # The num of threads to read tfrecords files.
  num_threads_processing_data = 16
  decode_output_speaker_volume_amp = False
  RESTORE_PHASE = 'GRIFFIN_LIM'  # 'MIXED','CLEANED','GRIFFIN_LIM'.
  GRIFFIN_ITERNUM = 50

  GPU_RAM_ALLOW_GROWTH = True
  USE_MULTIGPU = False  # dont't use multiGPU,because it is not work now...
  GPU_LIST = [0, 2]
  if USE_MULTIGPU:
    if batch_size % len(GPU_LIST) == 0:
      batch_size //= len(GPU_LIST)
    else:
      print('Batch_size %d cannot divided by gpu num %d.' %
            (batch_size, len(GPU_LIST)))
      exit(-1)

  minibatch_size = 400  # batch num to show

class C11_DATA:
  # rawdata, dirs by speakerid, like "....data_aishell/wav/train".
  #RAW_DATA = '/all_data/aishell2_speaker_list' # for docker
  RAW_DATA = '/home/room/work/lhf/alldata/aishell2_speaker_list'
  DATA_DICT_DIR = '_data/mixed_aishell'
  GENERATE_TFRECORD = True
  PROCESS_NUM_GENERATE_TFERCORD = 16
  TFRECORDS_NUM = 160  # 提多少，后面设置MAX_TFRECORD_FILES_USED表示用多少
  SHUFFLE = False

  LEN_WAWE_PAD_TO = 16000*3  # Mixed wave length (16000*3 is 3 seconds)
  UTT_SEG_FOR_MIX = [400, 460]  # Separate utt to [0:260],[260,290],[290:end]
  DATASET_NAMES = ['train', 'validation', 'test_cc']
  DATASET_SIZES = [600000, 18000, 100000]

  MAX_TFRECORD_FILES_USED=160 # <=TFRECORDS_NUM

  #NOISE_DIR = '/all_data/many_noise' # for docker
  NOISE_DIR = '/home/room/work/lhf/alldata/many_noise'
  NFFT = 512
  OVERLAP = NFFT - 256
  FS = 16000
  AUDIO_BITS = 16
  FEATURE_TYPE = 'LOG_MAG'  # MAG or LOG_MAG
  MASK_ON_MAG_EVEN_LOGMAG = True
  LOG_NORM_MAX = 6
  LOG_NORM_MIN = -0.3
  LOG_BIAS = 10000
  MAG_NORM_MAX = 1e6
  MAG_NORM_MIN = 0


  AUDIO_VOLUME_AMP=False


  MIX_METHOD = 'LINEAR' # "LINEAR" "SNR"
  MAX_SNR = 9  # 以不同信噪比混合
  MIN_SNR = -6
  #MIX_METHOD = "LINEAR"
  MAX_COEF = 1.0  # 以不同系数混合
  MIN_COEF = 0

  # TFRECORDS_DIR = '/all_data/feature_tfrecords' # for docker
  TFRECORDS_DIR = '/home/room/work/lhf/alldata/irm_data/feature_tfrecords_utt03s_irm'

NNET_PARAM = C11_NNET
MIXED_AISHELL_PARAM = C11_DATA
